### ISSUE TITLE
"payment amount must match shopping cart values" при использовании скидочного купона

### ISSUE THREAD

**Issue by qzya** [alexsaab/woocommerce_paymaster#1]
https://github.com/alexsaab/woocommerce_paymaster/blob/bd6675cd44aa29f7e38a36dea4761ff947e319ae/wp-content/plugins/paymaster-for-woocommerce/wc-paymaster.php#L268

Вот здесь, подозреваю, есть ошибка. В Вукоммерс есть еще такая штука, как скидочные купоны, и применяются они довольно хитро. Сумма купона делится на количество айтемов в корзине и частное от этого деления вычитается из _каждого_ айтема, и, соответственно, не всегда все делится на цело, а также результат зависит от округления и, по-видимому, настроек вукммерса: количества знаков после запятой. 
Но это все - сложный случай, нужно досконально разобраться как там все это происходит, вот же простой случай, из жизни:
Скажем, если изначальная сумма заказа = 1680 за 12 единиц по 140 рублей, сумма купона = 650, отнимаем от суммы 1680 купон 650, получаем 1030. Эта сумма будет у айтема в ['total']. Но если ее разделить на quantity т.е. на 12 (с двумя знаками после запятой), то получим 85.83 за единицу товара. И если обратно 85.83 умножим на 12, то получим 1029.96 - из-за этих 4 копеек paymaster ругается: **"payment amount must match shopping cart values"** - и не дает нашим клиентам оплатить покупки. Такие баги постоянно возникают на сайте. 
Очевидно делить нужно с опр. точностью, с той же, с которой делит woocommerce.

**Comment by alexsaab** [comment]
Да, конечно можно. Если не использовать купон, а использовать скидки в %%. А все купоны - зло. Пример 3 товара по 200 рублей. Купон на скидку, 100 рублей. Итого: 3*200-100 = 500 рублей, а теперь попробуйте без проблем  поделите на 3 и это только часть проблемы. Можно наверное как-то на кассе пробивать отрицательную сумму?! Что скажите?! Если да это решит 100% проблем.

С уважением, Алексей Агафонов

**Comment by qzya** [comment]
Согласен, но с этим злом надо как-то жить. 500 на 3 не поделится, а 499.98 - уже поделится. Можно проверять и корректировать сумму на стадии, где применяется купон, т.е. на чекауте и в корзине. Например, найти ближайшее число, меньшее 500, которое поделится на quantity и выписать и тут же применить "корректирующий" купон на 2 копейки.
